keras_cv/losses/giou_loss.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import tensorflow as tf
+
+from keras_cv import bounding_box
+
+
+class GIoULoss(tf.keras.losses.Loss):
+    """Implements the GIoU Loss
+
+    GIoU loss is a modified IoU loss commonly used for object detection. This loss aims
+    to directly optimize the IoU score between true boxes and predicted boxes. GIoU loss
+    adds a penalty term to the IoU loss that takes in account the area of the
+    smallest box enclosing both the boxes being considered for the iou. The length of
+    the last dimension should be atleast 4 to represent the bounding boxes. While
+    this dimension can have more than 4 values, these values will be ignored for the
+    calculation of this loss.
+
+    Args:
+        bounding_box_format: a case-insensitive string which is one of `"xyxy"`,
+            `"rel_xyxy"`, `"xyWH"`, `"center_xyWH"`, `"yxyx"`, `"rel_yxyx"`.
+            Each bounding box is defined by at least these 4 values. The inputs
+            may contain additional information such as classes and confidence after
+            these 4 values but these values will be ignored while calculating
+            this loss. For detailed information on the supported formats, see the
+            [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+
+    References:
+        - [GIoU paper](https://arxiv.org/pdf/1902.09630)
+
+    Sample Usage:
+    ```python
+    y_true = tf.random.uniform((5, 10, 5), minval=0, maxval=10, dtype=tf.dtypes.int32)
+    y_pred = tf.random.uniform((5, 10, 4), minval=0, maxval=10, dtype=tf.dtypes.int32)
+    loss = GIoULoss(bounding_box_format = "xyWH")
+    loss(y_true, y_pred).numpy()
+    ```
+    Usage with the `compile()` API:
+    ```python
+    model.compile(optimizer='adam', loss=keras_cv.losses.GIoULoss())
+    ```
+    """
+
+    def __init__(self, bounding_box_format, **kwargs):
+        super().__init__(**kwargs)
+        self.bounding_box_format = bounding_box_format
+
+    def _compute_giou(self, boxes1, boxes2):
+        boxes1_rank = len(boxes1.shape)
+        boxes2_rank = len(boxes2.shape)
+
+        if (
+            boxes1_rank != boxes2_rank
+            or boxes1_rank not in [2, 3]
+            or boxes2_rank not in [2, 3]
+        ):
+            raise ValueError(
+                "_compute_giou() expects both boxes to be batched, or both "
```

Comment:
instead of "_compute_giou() expects" tell users "`GIoULoss` expects"
tell users the public API name in value errors - thats who the errors are for in the end of the day